packages/synapse-react-client/src/components/DynamicForm/DynamicForm.tsx

```diff
@@ -0,0 +1,148 @@
+import React, { useState } from 'react'
+import { useQuery, useMutation } from '@tanstack/react-query'
+import Form from '@rjsf/mui'
+import validator from '@rjsf/validator-ajv8'
+import { displayToast } from '../ToastMessage'
+import { Box, Button, Typography } from '@mui/material'
+import { SkeletonParagraph } from '../Skeleton'
+
+export type DynamicFormProps = {
+  schemaUrl: string
+  uiSchemaUrl: string
+  postUrl: string
+  mutateFormDataBeforePost?: (formData: any) => any // allow caller to embed formdata into custom request body object
+  onCancel?: () => void
+}
+
+function DynamicForm(props: DynamicFormProps) {
+  const {
+    schemaUrl,
+    uiSchemaUrl,
+    postUrl,
+    mutateFormDataBeforePost,
+    onCancel,
+  } = props
+  const [formData, setFormData] = useState({})
+
+  // Fetch the form schema
+  const {
+    data: schema,
+    isLoading: _isSchemaLoading,
+    isError: isSchemaError,
+    error: schemaError,
+  } = useQuery({
+    queryKey: ['DynamicForm', 'schema', schemaUrl],
+    queryFn: async () => {
+      const response = await fetch(schemaUrl)
+      if (!response.ok) {
+        throw new Error('Error fetching schema')
+      }
+      return response.json()
+    },
+    enabled: !!schemaUrl,
+  })
+
+  // Fetch the UI schema
+  const {
+    data: uiSchema,
+    isLoading: _isUiSchemaLoading,
+    isError: isUiSchemaError,
+    error: uiSchemaError,
+  } = useQuery({
+    queryKey: ['DynamicForm', 'uiSchema', uiSchemaUrl],
+    queryFn: async () => {
+      const response = await fetch(uiSchemaUrl)
+      if (!response.ok) {
+        throw new Error('Error fetching UI schema')
+      }
+      return response.json()
+    },
+    enabled: !!uiSchemaUrl,
+  })
+
+  // Form submission mutation
+  const { mutate: submitFormData, isPending } = useMutation({
+    mutationFn: async (formData: any) => {
+      const formDataToSubmit = mutateFormDataBeforePost
+        ? mutateFormDataBeforePost(formData)
+        : formData
+      const response = await fetch(postUrl, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+          Accept: 'application/json',
+        },
+        body: JSON.stringify(formDataToSubmit),
+      })
+
+      if (!response.ok) {
+        const responseText = await response.text()
+        throw new Error(responseText)
+      }
+      return response.json()
+    },
+    onSuccess: () => {
+      displayToast('Form submitted successfully!', 'success')
+    },
+    onError: error => {
+      displayToast(`Error submitting form: ${error.message}`, 'danger')
+    },
+  })
+
+  const handleSubmit = (formData: any) => {
+    submitFormData(formData)
+  }
+
+  if (!schema || !uiSchema) {
```

Comment:
Need to explicitly check loading state. Since you're just checking the data, this will evaluate to true even if one or both queries are in an error state.
```suggestion
  if (_isSchemaLoading || _isUiSchemaLoading) {
```

Reply:
These are the loading flags that did not seem to be set properly.  Schema variables were undefined when these loading flags are false